Sources/MongoSwift/BSON/BSONEncoder.swift

```diff
@@ -293,7 +294,7 @@ extension _BSONEncoder {
         }
 
         // The top container should be a new container.
-        guard self.storage.count > depth else { return nil }
+        guard self.storage.count > depth else { return NSNull() }
```

Comment:
is this the right behavior here? similar checks in JSONEncoder seem to return either empty dictionaries or `nil`.

Reply:
So I thought this was correct since the returned result of `box()` is always added to `containers`, which should be real `BSONValue`'s.

Reply:
It may be clearer to ask what exactly this return value here is meant to indicate.

Is the return value from this line supposed to indicate:
1. A true null value?
2. A missing value?
3. An "error" or exceptional situation leading to a general Swift `nil`.

Tagging @kmahar, since she's most knowledgeable about Encoder/Decoder logic.

Reply:
so, this method is meant to be analogous to [`JSONEncoder`'s `box_` method.](https://github.com/apple/swift/blob/8b9cc180033d7c786a8646d7f71ae28802f69142/stdlib/public/SDK/Foundation/JSONEncoder.swift#L885) `box_` is called [at the very top level](https://github.com/apple/swift/blob/8b9cc180033d7c786a8646d7f71ae28802f69142/stdlib/public/SDK/Foundation/JSONEncoder.swift#L250) of `JSONEncoder.encode`, and a nil return value there throws, with the error message that the type created no top-level container - I think this would happen if you basically wrote an empty `encode(to:)` method for your type that never requested a container. so, I think a nil return here is meant to indicate that the type encoded *nothing*, not even an empty container.

Reply:
let's add some tests verifying our behavior matches JSONEncoder?

Reply:
Ah right, forgot about this. Thanks for bringing it up.

Reply:
Done, although would suggest you check it out in case I did something wrong with writing the test.